File: compiler/bench/npuzzle.c
// expected: 11111
// N-puzzle (8-puzzle) - count inversions to check solvability
#include <stdio.h>

#define SIZE 9

int puzzle[SIZE];

int count_inversions(void) {
    int count = 0;
    for (int i = 0; i < SIZE; i++) {
        for (int j = i + 1; j < SIZE; j++) {
            if (puzzle[i] > 0 && puzzle[j] > 0 && puzzle[i] > puzzle[j]) {
                count++;
            }
        }
    }
    return count;
}

int is_solvable(void) {
    return (count_inversions() & 1) == 0;
}

void init_puzzle(int seed) {
    for (int i = 0; i < SIZE; i++) puzzle[i] = i + 1;
    puzzle[SIZE - 1] = 0;

    int pos1 = seed % SIZE;
    int pos2 = (pos1 + 1) % SIZE;
    int tmp = puzzle[pos1];
    puzzle[pos1] = puzzle[pos2];
    puzzle[pos2] = tmp;
}

int main(void) {
    int count = 0;
    for (int i = 0; i < 100000; i++) {
        init_puzzle(i);
        if (is_solvable()) count++;
    }
    printf("%d\n", count);
    return 0;
}

File: compiler/bench/npuzzle.fs
\ expected: 11111
\ N-puzzle (8-puzzle) - count inversions to check solvability

9 constant SIZE
create puzzle SIZE cells allot

: p@ ( i -- val ) cells puzzle + @ ;
: p! ( val i -- ) cells puzzle + ! ;

: count-inversions ( -- count )
  0
  SIZE 0 do
    SIZE i 1+ do
      j p@ dup 0> if
        i p@ dup 0> if
          over > if swap 1+ swap then
        else drop then
      else drop then
    loop
  loop ;

: is-solvable? ( -- flag )
  count-inversions 1 and 0= ;

: init-puzzle ( seed -- )
  \ Simple scramble based on seed
  SIZE 0 do i 1+ i p! loop  \ 1 2 3 4 5 6 7 8 0
  0 SIZE 1- p!              \ put 0 at end

  \ Swap based on seed
  dup SIZE mod p@
  over SIZE mod
  over 1+ SIZE mod p@ swap
  2 pick SIZE mod p!
  swap 1+ SIZE mod p! drop ;

: main
  0
  100000 0 do
    i init-puzzle
    is-solvable? if 1+ then
  loop
  . cr ;
